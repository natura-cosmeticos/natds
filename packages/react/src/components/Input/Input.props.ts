import { IconName } from '@naturacosmeticos/natds-icons'
import { Size } from '@naturacosmeticos/natds-themes'

export type InputSize = keyof Pick<Size, 'medium' | 'mediumX'>
export type InputType = 'text' | 'password' | 'multiline'
export type Feedback = 'success' | 'error'

export interface BaseProps {
  /**
   * If `true`, the component is disabled.
   * @default false
   */
  disabled?: boolean,
  /**
   * The type of feedback to the user
   */
  feedback?: Feedback
  /**
   * The id of the `input` element, shared with label.
   */
  id: string
  /**
   * The helper text content.
   */
  helperText?: string
  /**
   * If filled, will appear above the field.
   */
  label?: string
  /**
   * Minimum number of rows to display when input type option is set to multiline.
   * @default 3
   */
  minRows?: number,
  /**
   * The onBlur attribute fires the moment that the element loses focus.
   */
  onBlur: React.FocusEventHandler<HTMLInputElement | HTMLTextAreaElement>,
  /**
   * The onChange event occurs when the value of an element has been changed.
   */
  onChange: React.ChangeEventHandler<HTMLInputElement | HTMLTextAreaElement>,
  /**
   * The onFocus event occurs when an element gets focus and is the opposite of the onblur event.
   */
  onFocus: React.FocusEventHandler<HTMLInputElement | HTMLTextAreaElement>
  /**
   * The short hint displayed in the `input` before the user enters a value.
   */
  placeholder?: string,
  /**
   * If `true`, the label is displayed as required with a `*` character
   * and the `input` element is required.
   */
  required?: boolean
  /**
   * If `true`, the user cannot change the value of the field
   * @default false
   */
  readOnly?: boolean,
  /**
   * If `true`, the user can expand the text field multiline area
   * @default false
   */
  isResizable?: boolean,
  /**
   * The height of the input
   * @default mediumX
   */
  size?: InputSize,
  /**
   * The type of the input
   * @default text
   */
  type?: InputType,
  /**
   * The value of the `input` element
   */
  value?: string,
}

export type InputActionIcon = BaseProps & ({
  /**
   * If `icon`, will display an Icon Button to the right of Input
   */
  action: 'icon'
  /**
   * Set the icon to be rendered.
   * Check all available names in [Icon Library](https://ds.natura.design/28db352be/p/94367e-icon-library/b/6154b9)
   * @default outlined-default-mockup
   */
  iconName: IconName
  /**
   * Use to define a text that explains the expected action
   */
  ariaLabel: string
  /**
   * Click handler
   */
  onClick: () => void
})

export type InputActionImage = BaseProps & ({
  /**
   * If `image`, will display an image to the right of Input
   */
  action: 'image'
  /**
   * Provides alternative information for an image if a user for some reason cannot view it
   */
  alt: string
  /**
   * The path for the image
   */
  src: string
})

export type InputProps = BaseProps & { action?: never }
  | InputActionIcon
  | InputActionImage
