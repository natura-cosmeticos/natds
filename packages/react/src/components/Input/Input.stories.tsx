import React from 'react'
import { Story, Meta } from '@storybook/react'
import { Input, InputProps } from '.'

export default {
  title: 'Utilities/Input',
  component: Input,
  parameters: {
    previewTabs: {
      canvas: { hidden: true }
    }
  }
} as Meta

export const Default: Story<InputProps> = (args) => <Input {...args} />
Default.args = {
}
