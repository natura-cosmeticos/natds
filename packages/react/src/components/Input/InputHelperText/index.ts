import InputHelperText from './InputHelperText'

export { InputHelperText }
export default InputHelperText
