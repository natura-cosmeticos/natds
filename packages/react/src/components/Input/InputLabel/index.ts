import InputLabel from './InputLabel'

export { InputLabel }
export default InputLabel
