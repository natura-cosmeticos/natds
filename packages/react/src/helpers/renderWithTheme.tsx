import React from 'react'
import { JssProvider, SheetsRegistry } from 'react-jss'
import { render } from '@testing-library/react'
import { ThemeProvider } from '../ThemeProvider'

const renderWithTheme = (component) => {
  const sheets = new SheetsRegistry()

  const renderTree = render(
    <JssProvider registry={sheets}>
      <ThemeProvider>
        {component}
      </ThemeProvider>
    </JssProvider>
  )

  return { styles: sheets.toString(), component: renderTree }
}

export default renderWithTheme
