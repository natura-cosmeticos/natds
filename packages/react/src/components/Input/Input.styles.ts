/* eslint-disable max-lines-per-function */
import { createUseStyles } from 'react-jss'
import { Theme } from '@naturacosmeticos/natds-themes'
import { InputProps } from './Input.props'

type InputStyleProps = Pick<InputProps, 'feedback' | 'disabled' | 'readOnly'| 'size' | 'action' | 'isResizable'>

const getFeedbackBorderColor = (theme: Theme) => ({ feedback }: InputStyleProps) => {
  const borderColor = {
    error: theme.color.alert,
    success: theme.color.success
  }

  return feedback ? borderColor[feedback] : theme.color.lowEmphasis
}

const styles = createUseStyles((theme: Theme) => ({
  wrapper: {
    alignItems: 'center',
    backgroundColor: theme.color.surface,
    border: '1px solid',
    borderColor: getFeedbackBorderColor(theme),
    borderRadius: theme.textField.borderRadius,
    cursor: ({ disabled, readOnly }: InputStyleProps) => (!disabled && !readOnly ? 'text' : 'default'),
    display: 'flex',
    justifyContent: 'space-between',
    position: 'relative',
    '&:hover': {
      borderColor: ({ disabled, readOnly }: InputStyleProps) => (
        !disabled && !readOnly && theme.color.mediumEmphasis
      )
    },
    '&:focus-within': {
      border: '2px solid',
      borderColor: theme.color.primary
    },
    '&:after': {
      backgroundColor: theme.color.lowEmphasis,
      content: '""',
      height: '100%',
      left: 0,
      opacity: ({ readOnly }: InputStyleProps) => (readOnly ? theme.opacity.disabledLow : 0),
      pointerEvents: 'none',
      position: 'absolute',
      top: 0,
      width: '100%'
    }
  },
  base: {
    background: 'none',
    border: 'none',
    color: theme.color.highEmphasis,
    fontFamily: [
      theme.textField.content.primary.fontFamily,
      theme.textField.content.fallback.fontFamily
    ],
    fontSize: theme.textField.content.fontSize,
    fontWeight: theme.textField.content.primary.fontWeight,
    letterSpacing: theme.textField.content.letterSpacing,
    lineHeight: theme.textField.content.lineHeight,
    outline: 'none',
    padding: theme.spacing.small,
    width: '100%',
    '&:read-only': {
      pointerEvents: 'none'
    },
    '&:disabled': {
      color: theme.color.lowEmphasis,
      cursor: 'default'
    }
  },
  input: {
    extend: 'base',
    boxSizing: 'border-box',
    height: ({ size }: InputStyleProps) => size && theme.size[size],
    paddingRight: ({ action }: InputStyleProps) => action && theme.spacing.tiny,
    '&::placeholder': {
      color: ({ disabled }: InputStyleProps) => (
        disabled ? theme.color.lowEmphasis : theme.color.mediumEmphasis
      )
    }
  },
  textArea: {
    extend: 'base',
    paddingRight: ({ action }: InputStyleProps) => action && theme.spacing.tiny,
    resize: ({ isResizable }: InputStyleProps) => !isResizable && 'none',
    '&::placeholder': {
      color: ({ disabled }: InputStyleProps) => (
        disabled ? theme.color.lowEmphasis : theme.color.mediumEmphasis
      )
    }
  }
}))

export default styles
