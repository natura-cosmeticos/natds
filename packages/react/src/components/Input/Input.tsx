/* eslint-disable max-lines-per-function */
import React from 'react'
import { InputProps } from './Input.props'
import styles from './Input.styles'
import InputAction from './InputAction'

const Input = (props: InputProps): JSX.Element => {
  const {
    disabled = false,
    id,
    isResizable = false,
    minRows = 3,
    onBlur,
    onChange,
    onFocus,
    placeholder,
    readOnly = false,
    size = 'mediumX',
    type = 'text',
    feedback,
    value,
    ...rest
  } = props

  const classes = styles({
    isResizable, size, feedback, ...props
  })
  const isMultiline = type === 'multiline'

  return (
    <div className={classes.wrapper}>
      {
        isMultiline
          ? (
            <textarea
              className={classes.textArea}
              data-testid="ds-input-multiline"
              disabled={disabled}
              onBlur={onBlur}
              onChange={onChange}
              onFocus={onFocus}
              placeholder={placeholder}
              readOnly={readOnly}
              rows={minRows}
              id={id}
            />
          )
          : (
            <input
              className={classes.input}
              data-testid="ds-input"
              disabled={disabled}
              onBlur={onBlur}
              onChange={onChange}
              onFocus={onFocus}
              placeholder={placeholder}
              readOnly={readOnly}
              type={type}
              value={value}
              id={id}
            />
          )
      }
      {rest.action && (<InputAction {...props} />)}
    </div>
  )
}

export default Input
