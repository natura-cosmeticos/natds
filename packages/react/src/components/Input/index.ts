import Input from './Input'

export { Input }
export { InputProps } from './Input.props'
export default Input
