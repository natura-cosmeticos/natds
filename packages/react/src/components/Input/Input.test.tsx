import React from 'react'
import Input from '.'
import renderWithTheme from '../../helpers/renderWithTheme'
import { InputProps } from './Input.props'

const defaultProps: InputProps = {
  onBlur: () => '',
  onChange: () => '',
  onFocus: () => ''
}

describe('Input component', () => {
  it('should render correctly with default props', () => {
    const { styles, component } = renderWithTheme(<Input {...defaultProps} testId="input-text" />)

    expect([styles.toString(), component.container]).toMatchSnapshot()
    expect(component.getByTestId('input-text')).not.toBeDisabled()
    expect(component.getByTestId('input-text')).not.toBeRequired()
  })

  it('should render correctly when type is multiline', () => {
    const { styles, component } = renderWithTheme(<Input {...defaultProps} type="multiline" testId="input-multiline" />)

    expect([styles.toString(), component.container]).toMatchSnapshot()
    expect(component.getByTestId('input-multiline')).not.toBeDisabled()
    expect(component.getByTestId('input-multiline')).not.toBeRequired()
    expect(component.getByTestId('input-multiline')).toHaveProperty('readOnly', false)
  })
})
