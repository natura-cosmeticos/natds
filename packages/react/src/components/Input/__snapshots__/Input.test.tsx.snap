// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should render correctly when type is multiline 1`] = `
Array [
  ".wrapper-0-2-1 {
  border: 1px solid;
  display: flex;
  position: relative;
  align-items: center;
  border-radius: 4px;
  justify-content: space-between;
  background-color: #ffffff;
}
.wrapper-0-2-1:hover {}
.wrapper-0-2-1:focus-within {
  border: 2px solid;
  border-color: #f4ab34;
}
.wrapper-0-2-1:after {
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  content: \\"\\";
  position: absolute;
  pointer-events: none;
  background-color: #bbbbbb;
}
.base-0-2-2 {
  color: #333333;
  width: 100%;
  border: none;
  outline: none;
  padding: 16px;
  font-size: 16px;
  background: none;
  font-family: Roboto, Arial;
  font-weight: 400;
  line-height: 1.25;
  letter-spacing: 0.51px;
}
.base-0-2-2:read-only {
  pointer-events: none;
}
.base-0-2-2:disabled {
  color: #bbbbbb;
  cursor: default;
}
.input-0-2-3 {
  color: #333333;
  width: 100%;
  border: none;
  outline: none;
  padding: 16px;
  font-size: 16px;
  background: none;
  box-sizing: border-box;
  font-family: Roboto, Arial;
  font-weight: 400;
  line-height: 1.25;
  letter-spacing: 0.51px;
}
.input-0-2-3:read-only {
  pointer-events: none;
}
.input-0-2-3:disabled {
  color: #bbbbbb;
  cursor: default;
}
.input-0-2-3::placeholder {}
.textArea-0-2-4 {
  color: #333333;
  width: 100%;
  border: none;
  outline: none;
  padding: 16px;
  font-size: 16px;
  background: none;
  font-family: Roboto, Arial;
  font-weight: 400;
  line-height: 1.25;
  letter-spacing: 0.51px;
}
.textArea-0-2-4:read-only {
  pointer-events: none;
}
.textArea-0-2-4:disabled {
  color: #bbbbbb;
  cursor: default;
}
.textArea-0-2-4::placeholder {}
.wrapper-d0-0-2-5 {
  border-color: #bbbbbb;
  cursor: text;
}
.wrapper-d0-0-2-5:hover {
  border-color: #777777;
}
.wrapper-d0-0-2-5:after {
  opacity: 0;
}
.input-d1-0-2-6 {
  height: 56px;
}
.input-d1-0-2-6::placeholder {
  color: #777777;
}
.textArea-d2-0-2-7 {
  resize: none;
}
.textArea-d2-0-2-7::placeholder {
  color: #777777;
}",
  <div>
    <div
      class="wrapper-0-2-1 wrapper-d0-0-2-5"
    >
      <textarea
        class=" textArea-0-2-4 textArea-d2-0-2-7"
        data-testid="ds-input-multiline"
        id="example_to_test"
        rows="3"
      />
    </div>
  </div>,
]
`;

exports[`Input component should render correctly with default props 1`] = `
Array [
  ".wrapper-0-2-1 {
  border: 1px solid;
  display: flex;
  position: relative;
  align-items: center;
  border-radius: 4px;
  justify-content: space-between;
  background-color: #ffffff;
}
.wrapper-0-2-1:hover {}
.wrapper-0-2-1:focus-within {
  border: 2px solid;
  border-color: #f4ab34;
}
.wrapper-0-2-1:after {
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  content: \\"\\";
  position: absolute;
  pointer-events: none;
  background-color: #bbbbbb;
}
.base-0-2-2 {
  color: #333333;
  width: 100%;
  border: none;
  outline: none;
  padding: 16px;
  font-size: 16px;
  background: none;
  font-family: Roboto, Arial;
  font-weight: 400;
  line-height: 1.25;
  letter-spacing: 0.51px;
}
.base-0-2-2:read-only {
  pointer-events: none;
}
.base-0-2-2:disabled {
  color: #bbbbbb;
  cursor: default;
}
.input-0-2-3 {
  color: #333333;
  width: 100%;
  border: none;
  outline: none;
  padding: 16px;
  font-size: 16px;
  background: none;
  box-sizing: border-box;
  font-family: Roboto, Arial;
  font-weight: 400;
  line-height: 1.25;
  letter-spacing: 0.51px;
}
.input-0-2-3:read-only {
  pointer-events: none;
}
.input-0-2-3:disabled {
  color: #bbbbbb;
  cursor: default;
}
.input-0-2-3::placeholder {}
.textArea-0-2-4 {
  color: #333333;
  width: 100%;
  border: none;
  outline: none;
  padding: 16px;
  font-size: 16px;
  background: none;
  font-family: Roboto, Arial;
  font-weight: 400;
  line-height: 1.25;
  letter-spacing: 0.51px;
}
.textArea-0-2-4:read-only {
  pointer-events: none;
}
.textArea-0-2-4:disabled {
  color: #bbbbbb;
  cursor: default;
}
.textArea-0-2-4::placeholder {}
.wrapper-d0-0-2-5 {
  border-color: #bbbbbb;
  cursor: text;
}
.wrapper-d0-0-2-5:hover {
  border-color: #777777;
}
.wrapper-d0-0-2-5:after {
  opacity: 0;
}
.input-d1-0-2-6 {
  height: 56px;
}
.input-d1-0-2-6::placeholder {
  color: #777777;
}
.textArea-d2-0-2-7 {
  resize: none;
}
.textArea-d2-0-2-7::placeholder {
  color: #777777;
}",
  <div>
    <div
      class="wrapper-0-2-1 wrapper-d0-0-2-5"
    >
      <input
        class=" input-0-2-3 input-d1-0-2-6"
        data-testid="ds-input"
        id="example_to_test"
        type="text"
        value=""
      />
    </div>
  </div>,
]
`;
